Converge IMU faster at higher groundspeed during GPS Rescue

ctzsnooze opened the pull request:
Tested, works well, but would appreciate further testing please!  

Makes the following changes:

- Yaw IMU gain increases up to 6x when there is a large drift over ground error, so that the IMU can re-orient itself more quickly the worse the error
- locks yaw IMU gain to zero during climb and rotate phases, preventing acquisition of IMU orientation errors due to drift during long climbs
- attenuates the maximum angle allowed when there is a large drift over ground error, to reduce the speed of the flyaway that can happen when the IMU orientation is very incorrect
- max rescue angle defaults to 45 degrees.  Even with a maximum IMU error, 45 degrees max angle doesn't cause the quad to speed off madly into the distance.
- fixes an issue where direct entry to a descent mode may not gain full yaw authority
- changes the sensitivity of the `FLY HOME` sanity check.  Now the sanity check will only count up towards to the 15s disarm value if the quad fails to reach 10%, not 50%, of the set return speed, when checked each second.  As before this is an up-down counter, so any time the quad attains >10% of the set return velocity to home, the counter decrements by 1, towards a minimum of zero, and when less than 10%, the counter increments by one.  If there is a net cumulative period of more than 15s with less than 10% forward speed to home, the sanity check will disarm the quad, since it is most unlikely that it will return home.  Users should expect some counting up in the early part of a rescue, especially with bad IMU disorientation, but with the above changes the quad should not disarm from this sanity check unless something really has gone badly wrong, eg stuck in a tree, or crashed, or extremely high wind.

WARNING 1: 
A GPS-only rescue system cannot handle high wind drift - do NOT rely on GPS Rescue in winds above 50-60kph unless you have a properly calibrated and fully functional, tested, Magnetometer (compass).  Higher maximum angle will help handle strong winds better, but the max throttle must be increased also, and in the face of a temporary IMU error, the quad may fly off in the wrong direction at high speed.  In practice, a 35-45 degree angle is more than enough for normal conditions, a 50-55 degree angle may be better in stronger winds.  Always test the return in the conditions of the day to ensure your quad will turn and head for home effectively.  If the wind is too high, the quad will kind of spiral around, at speed, making little or no headway.  Don't let it do this for more than 15s or it will sanity check and drop.

WARNING 2:
IMU disorientation is always possible with GPS-only systems. If a rescue is initiated and the IMU is disoriented from prior drift or high wind, the quad may initially head off in completely the wrong direction, even though the arrow points straight up.  Functionally the quad thinks it is flying home, but has temporarily got the orientation of the quad wrong.  The IMU will, or should, correct itself over the next 2-3seconds, but will fly in a large arc while doing so.  If the maxim allows angle in the rescue is set high, this arc may be flown at high speed, and cover some considerable distance, but the quad will likely handle higher wind rescues a bit more reliably.

To test: enter a GPS Rescue after a 6-8s period of tail drift (fly backwards at > 2m/s for say 6-10s).  Hold that drift long enough for the arrow to rotate 180 degrees wrong.  Initiate the rescue. The rescue will then rotate entirely the wrong direction, ie fly directly away from home.  With this PR it should gain speed away from home for less than a second, and then smoothly and steadily, over 2-3 seconds rotate towards home.  The arrow will keep pointing straight up all the time (the turning is the IMU correcting itself).  After maybe 2-4s, it should be heading home.  The quad will attain a decent speed while correcting itself.  The max speed is determined mostly by the maximum angle. 

If the debug mode `RTH` is displayed on the OSD screen, the sanity check timer is the value of the last two digits of the last debug element.

Testing in high wind and/or with prior extreme back drift would be appreciated.
Testing to figure out how much wind can be tolerated would be helpful, too.

Thanks to @Sek101 for investigating and clarifying the problem.  To summarise, the problem is that:

- if the quad drifts in a direction other than in the direction of the nose of the quad, and
- the drift is at a significant speed (anything > 2m/s), and
- the drift is sustained for at least one or two dcm_kp time constants (6-8 seconds)
- the IMU will, over time, consider that the nose of the quad is pointing in the direction of the course over ground, not the true heading of the quad, and point the arrow to the bottom of the goggles, thinking that is where home is.
- Then, when a rescue is initiated, the quad will rotate so that the arrow points upwards, and will fly in a direction that can be entirely incorrect, including a full 180 degrees wrong.  It will also fly faster than usual at first, and may make a large arc in the sky before finally settling on the true path to home.

This excellent video from Sek101 shows a takeoff, then a quick rotation to fly away from home backwards.  The home arrow slowly orients itself to become 180 degrees wrong.  When the rescue is initiated, the quad rotates according to the incorrect IMU orientation, and flies off 180 degrees in the wrong direction.  
https://www.youtube.com/watch?v=DyEYdI-Qgtw

Code attempt 1 = FAILURE!
This idea was simple: make the dcm_kp value proportional to groundspeed during a GPS Rescue, using about a 4x increase in dcm_kp at normal rescue speed, and very low dcm_kp at low/zero groundspeed.  
RESULT: spirals during initiation when IMU error existed at the start, quad flies off very fast, then enters a kind of "chasing the tail" behaviour.
Video shows, around 44s, quad in reverse, arrow going wrong, then persisting rotation.
https://www.youtube.com/watch?v=BSOgY-f1XtM

Code attempt 2 = NOT QUITE THERE YET!
Here we detect an absolute error between groundspeed and velocity to home, normalised to the intended return speed, plus one.  If the quad is flying directly home, there is no velocity error, so this `groundspeedErrorRatio` value would be 0+1 , ie  1.0.  If the quad is flying off at 90 degrees to home at a groundspeed equal to the intended return speed, the `groundspeedErrorRatio` would be 2.0.  If flying off with a full 180 degree error, at the intended return speed, this factor would be 3.0.  
Then we use this factor to boost `dcm_kp` during the rescue.
RESULT: Some improvement, but still tended to spiral and pick up a lot of speed.

Code attempt 3 = very good!
- calculate `groundspeedErrorRatio` as above, and use it to boost `dcm_kp` up to 5 times.
- use the inverse of this value to attenuate the maximum allowed pitch angle, thereby limiting absolute velocity when the IMU error would send the quad in the wrong direction.
- set `dcm_kp` forcibly to zero while in the climb or rotate phases of the rescue.  This locks in the pre-existing IMU state, which is normally pitch forward flight with a high likelihood of pointing to home.  By locking in that IMU state, we prevent drift (from wind, etc) from building up an IMU error of the kind that could cause a flyaway.
- finally, restore the max angle to the previous default of 40, for pilots who may have GPS modules that aren't working well, to avoid excessive pitch error issues in those cases.

Here is some low-res video of the IMU problem and how this PR achieves a pretty good result in the face of the worst possible IMU errors.

https://www.youtube.com/watch?v=IKbeWCsQj24

This PR should significantly reduce the chance that a working GPS Rescue setup would fail, especially if there is, or has been, signficant sideways or rear-wards drift either for a period of time before the rescue was initiated, or during the climb period.  Previous versions of GPS Rescue were very susceptible to IMU disorientation and could enter persistent non-recovering flyaways.

ledvinap commented:
(Did I mention Kalman filter in IMU this year?)

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> FAIL
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

KarateBrot commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +569,18 @@ static void sensorUpdate(void)
+        // mulitply by 2 means if the groundspeed errorRatio is 1, dcm_kp will be 2x normal; if error is 2, then 4x dcm_kp, max 10x
+    }
+    
+    
```
```suggestion
```

KarateBrot commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
+    if (gpsRescueConfig()->rescueGroundspeed) {
+        const float setGroundspeedInv = 1.0f / gpsRescueConfig()->rescueGroundspeed;
+        float groundspeedErrorRatio = fabsf(rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) * setGroundspeedInv;
+        // zero if groundspeed = velocity to home,
```
```suggestion
        // 0 if groundspeed = velocity to home,
```

ctzsnooze replied:
thanks, done.

KarateBrot commented on src/main/flight/gps_rescue.c:
```diff
@@ -535,7 +538,7 @@ static void sensorUpdate(void)
     DEBUG_SET(DEBUG_GPS_RESCUE_TRACKING, 2, lrintf(rescueState.sensor.currentAltitudeCm));
     DEBUG_SET(DEBUG_GPS_RESCUE_THROTTLE_PID, 2, lrintf(rescueState.sensor.currentAltitudeCm));
-    DEBUG_SET(DEBUG_GPS_RESCUE_HEADING, 0, rescueState.sensor.groundSpeedCmS); // groundspeed cm/s
+//    DEBUG_SET(DEBUG_GPS_RESCUE_HEADING, 0, rescueState.sensor.groundSpeedCmS); // groundspeed cm/s
```
```suggestion
    DEBUG_SET(DEBUG_GPS_RESCUE_HEADING, 0, rescueState.sensor.groundSpeedCmS); // groundspeed cm/s
```

ctzsnooze replied:
thanks, sorry to have missed that.  I 'hijacked' that debug for testing purposes and forgot to restore it.  Much appreciated!

jsk2084 commented:
Tested and found this PR to be an excellent solution to the heading error problem that can sometimes arise when not using a compass. The following clip starts with a quad flying around home point, drifting sideways then inducing full heading error by flying backwards for some time before starting a Rescue. After climb, it quickly and neatly corrects it's course towards home! Also tested with a max angle setting of 60 degrees and groundspeed of 50kph... No issues encountered so far! https://youtu.be/6MVMsMArtVo

ctzsnooze commented:
Added a small change to the `FLY HOME` phase sanity check to only count up when return velocity to home is <25% of set velocity to home, and (sorry for going up and down recently) set the default angle to 45 degrees.  These changes should make it less likely to trigger a sanity in strong headwinds that could slow the quad down.

If, for instance, the quad travels at 80kph at 45 degree pitch angle, a headwind of 60kph would reduce net speed to home to only 20kph, or into the range that would cause a 25% sanity check threshold to fail.  Users who want to fly long distances with strong tail winds should consider a higher GPS max angle value, and consider the possible time taken to fly home against such a wind, since throttle may be high for a long time, and the battery may get very low.

Thanks to @jsk2084 for confirming that 60 degrees is OK with the changes in this PR.  Something around 70 degrees would seem to be the practical absolute upper limit.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -758,7 +777,7 @@ void gpsRescueUpdate(void)
         }
         if (rescueState.sensor.absErrorAngle < 30.0f) {
-            rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle; // allow pitch
+            rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.sensor.groundspeedErrorRatioInv; // allow pitch
```
Is this really necessary? IMO it will be very confusing for someone reading the code.
At least use `groundspeedErrorRatio` and comment its range/use here

Also absErrorAngle < 30, so `groundspeedErrorRatio` would be quite small (groundspeedErrorRatio is function of absErrorAngle )

ctzsnooze replied:
The absErrorAngle is a heading error, but if the IMU is wrong, the heading error can be tiny, whereas the groundspeed error ratio can be very high.

For example, when the IMU is 180 degrees wrong, as in the example above, and if the quad is drifting away quite fast, the groundspeed error ratio can be say 3 or 4.  Like in the example videos above.  

In this situation, when the quad rotates after the climb, the rotation continues until the heading error is brought to zero.  

Howeve, due to the IMU error, the true heading of the quad is 180 degrees wrong when the measured heading error is zero.  After rotating so that the heading error is 30 degrees (which is the point at which we start to pitch forward), a quad in this situation is pointing 150 degrees away from home.  Over the next half a second, that remaining 30 degrees of error is removed, and then the nose is pointing a full 180 degrees from home.  

In this example, we then have a groundspeedErrorRatio of 3 or 4, even though the heading error is close to zero.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -792,10 +811,13 @@ void gpsRescueUpdate(void)
-
+ 
         if (newGPSData) {
+            rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.sensor.groundspeedErrorRatioInv;
```
Better to use `else` branch (set `rescueState` only in one place)

ctzsnooze replied:
I'm not sure how to avoid this... in many places in the rescue code, these rescueState values are changed sequentially as we pass through the stages of the rescue, and used elsewhere. They are usually only changed at one time, in one place, sequentially, and this is no exception.

ledvinap reviewed (commented):
Personally, I don't like heuristics without physical background. It is probably working, but it will pile up in the code, eventually making the code impossible to understand.

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -535,7 +538,7 @@ static void sensorUpdate(void)
     DEBUG_SET(DEBUG_GPS_RESCUE_TRACKING, 2, lrintf(rescueState.sensor.currentAltitudeCm));
     DEBUG_SET(DEBUG_GPS_RESCUE_THROTTLE_PID, 2, lrintf(rescueState.sensor.currentAltitudeCm));
-    DEBUG_SET(DEBUG_GPS_RESCUE_HEADING, 0, rescueState.sensor.groundSpeedCmS); // groundspeed cm/s
+//    DEBUG_SET(DEBUG_GPS_RESCUE_HEADING, 0, rescueState.sensor.groundSpeedCmS); // groundspeed cm/s
```
thanks, have fixed this, it was a left over from testing

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
 
+    // values to modify dcm_kp to a low value when climbing, and a higher value when there is a velocity error
+    if (gpsRescueConfig()->rescueGroundspeed) {
+        const float setGroundspeedInv = 1.0f / gpsRescueConfig()->rescueGroundspeed;
+        float groundspeedErrorRatio = fabsf(rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) * setGroundspeedInv;
```
It's true that a cosine would be better, but cos is a lot more expensive to compute, and the application here doesn't require that level of precision to effectively solve the problem.

`fobsf` is intended.  It returns the absolute error between the velocity to home and the groundspeed.  Groundspeed is always positive, but velocity to home can be negative.  It is very unusual for the velocity to home to be greater than the set groundspeed, since when the IMU is oriented accurately, the two are very close to being identical, and when the IMU is oriented incorrectly, the velocity to home is almost always less than the groundspeed, in the rescue context.  The intent is to return a value such that when the velocity to home and the groundspeed are the same, we return 0; when they differ, we return the magnitude of the difference as a positive number, expressed as a fraction of the set groundspeed.

The `groundspeedErrorRatio` value then represents absolute error between the velocity to home, and groundspeed, expressed as a fraction of the set return speed, with the lowest possible value being 1.  When there is no error, the value is 1; and it increases as the error increases, to a max of 5.  It is a simple scaled velocity to home error indicator.  

The 'normal' situation would be quad flying to home, with the IMU oriented in the pitch angle.  The velocity to home would equal the GPS groudspeed, and both would be equal to, or less than, the set groundspeed.  In this case, the `groundspeedErrorRatio` is very close to 1.0.  As a result, dcm_kp is unchanged, and the pitch angle limit is normal.  The result is the normal pitch response while the IMU is oriented correctly, and an IMU that adapts heading relatively slowly.

Let's imagine a rescue like in the video above where the IMU has oriented itself 90 degrees wrong.  The rescue has started, the speed over ground is stable at the set return velocity, but the quad is flying sideways, at 90 degrees relative to home, with zero velocity to home.  The error factor would then be 2.0 (1.0 + (1-0)/1).  This can only can only be a stable condition in a rescue when the IMU is wrong.  

Without this PR, a quad in this situation would fly sideways and take maybe 6 seconds to correct the IMU.  The PIDs target velocity to home, the quad would probably assume a high angle, and gain a high velocity, still at 90 degrees to home, taking it even further from home.  By the time the IMU has corrected itself, the quad is likely to be a long way from home, but with a bit of luck it will eventually get home.

With this PR, we use the  the `groundspeedErrorRatio` to boost dcm_kp by a factor of 2, which halves the time it takes the IMU to correct itself.  We also use the inverse of the `groundspeedErrorRatio` to cut the max allowed pitch angle by 50%, reducing forward velocity.  Should that be insufficient, and should the quad speed up even more, the dcm_kp value will increase more, and the maximum pitch angle limit will get smaller.  The result is a kind of speed limiter.  We get some speedup, but not an excessive amount.  The IMU corrects much more quickly, and without speeding off so fast, the quad rotates to home with a much tighter radius.  This can be seen in the videos above.

As the quad rotates to point more truly to home, the `groundspeedErrorRatio` falls.  This allows the quad to make a higher pitch angle if it needs to, and the dmc_kp value eases back towards normal.  After a short time we are back in the normal situation, as above, with groundspeed and velocity to home being comparable, and the velocity accurately controlled by the PIDs.

An even more difficult situation is where there is an IMU error of 180 degrees and the quad is flying 180 degrees away from home.

Imagine it has prior drift at 3x the set groundspeed, away from home, when the rescue starts.  The `groundspeedErrorRatio` would be 4.0 in this case.  This indicates a very large error between the IMU orientation and the true orientation of the quad.  We use this high error value to boost dcm_kp by four times, reducing its time constant, or lag, by a factor of 4, leading to much more rapid correction than would otherwise occur, reducing the amount of time spent flying the wrong way, and tightening the radius of the correcting turn.  At the same time, the max pitch angle is constrained to only 25% of normal, which reduces the speed away from home considerably, tightening the radius of the turn and limiting the distance traveled away.  Instead of flying hundreds of meters away and potentially failing the sanity check, the quad turns relatively quickly and gets back on track.  This is seen in the videos above.

That's the intent of the code.  There is an arbitrary element to it, I accept.  At the same time, it makes sense, and is relatively simple.

I have grappled with this for quite a long time, and was unable to find a better solution.  I explored solutions based on IMU vs GPS heading differences, simple velocity limiters, and other methods, and none were as effective, and most caused problems in edge case scenarios.

At some point, when we have position control and the ability to use position control to fly between known waypoints, all this code should be replaced.  But for now, this is the best solution I can think of, and it really does solve a problem that has been around for a very, very long time, in a reasonably simple manner.

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
+        // zero if groundspeed = velocity to home,
+        // 1 if flying sideways at target groundspeed but zero home velocity,
+        // 2 if flying backwards at target groundspeed
+        groundspeedErrorRatio = constrainf(1.0f + groundspeedErrorRatio, 1.0f, 5.0f);
```
The comments above refer to `groundspeedErrorRatio`, calculated on the immediate line above the comments.
I then add 1 to `groundspeedErrorRatio` and constrain it to the range 1 - 5.

ledvinap replied:
That's the problem - groundspeedErrorRatio above, groundspeedErrorRatio below ..
Either use another variable (compiler will optimize it away), or add +1 in first equation

ctzsnooze replied:
I revised the code just now, I think the confusion is gone.

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
+        float groundspeedErrorRatio = fabsf(rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) * setGroundspeedInv;
+        // zero if groundspeed = velocity to home,
+        // 1 if flying sideways at target groundspeed but zero home velocity,
+        // 2 if flying backwards at target groundspeed
```
Probably more accurate, in a general sense, to say 'moving away from home'.  I'll do that.
You're right, the nose could be pointing towards home, or away from home (or any direction at all, but let's consider the two extremes).
If the nose is pointing towards home, pitching forward will work normally, accelerating towards to home, reducing the negative velocity to home.  
If the nose is pointing away from home, this can only be the case, after rotation is complete, if the IMU is 180 degrees wrong.  Pitching forward with twill just make the quad move faster away from home.  We need to correct the IMU error quickly, and to limit the pitch angle to minimise speed away from home.  The PR achieves that.

ledvinap replied:
(It is possible that IMU handles 180deg case very poorly). 
Also it would make sense to only increase GPS heading gain, attitude is probably fine.

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
+        // limit to max 5
+        const bool climbOrRotate = (rescueState.phase == RESCUE_ATTAIN_ALT) || (rescueState.phase == RESCUE_ROTATE);
+        rescueState.sensor.dcm_kpModifier = (climbOrRotate) ? 0.0f : groundspeedErrorRatio;
+        // zero dcm_kp during climb or rotate, to stop IMU error accumulation arising from drift during long climbs
```
Yes, I think I put examples or explanations after the line, depending on context.  I like to read the code first, then I can ignore the comment, if I can follow the code and jump visually to the next line of code.  In this case if it's OK I'd prefer to leave it as is.

ctzsnooze commented:
@ledvinap thanks for your comments, I've implemented what I can.  The problem I seek to solve arises only when there is IMU disorientation, and this happens in only a very small proportion of rescues, typically those with long climb times in high wind.  There have been a number of reports and in all cases drift during the climb eventually led me to an understanding of how things went wrong.  It has been a really tough challenge to find a solution. This solution definitely works better than anything else I could come up with.  I wish we had a better solution.  If anyone can find one, that would be great.  But for now this 'works'.

ledvinap commented:
Maybe there should be 'IMU heading align' GPS phase after climbing to altitude.
IMU attitude shall be quite fine after climb (maybe stay in climb phase for few seconds it is too short).
Set forward pitch limit to some reasonable value and speed setpoint to value where there is good GPS heading acquisition. Wait until GPS and IMU headings match, with RTH heading PID active (to track decreasing IMU error). Increase gain here if necessary.
Then enter normal RTH phase (probably speeding up to full RTH speed).

ledvinap commented:
Also, CoG compensation is probably very poor when error angle is 180 deg. https://github.com/betaflight/betaflight/blob/master/src/main/flight/imu.c#L222 

IIRC it is calculating vector cross product, which is small for antiparallel case. It may be special-cased for error >~ 90deg, applying constant 1 instead of sin(theta) in that case.

ctzsnooze commented:
> Maybe there should be 'IMU heading align' GPS phase after climbing to altitude.

@ledvinap it would be very nice if we could know for sure that the IMU was good, but in cases of strong wind, or a poorly trimmed accelerometer, the climb phase course over ground could be very different from the direction to home, and the IMU will then assume that the nose is pointing in the direction of the course over ground, and become very incorrect, especially for long climbs which give it plenty of time.

That's why I lock the IMU during climbs.  As you appreciate, it would be best to lock it only for yaw, I would like that better than the current situation.

In strong wind, at the end of the climb, the quad could be blown backwards at high speed.  This is an essentially impossible case to manage once we complete the climb.  If we apply a certain amount of pitch, and if that amount is inadequate to overcome the wind, even when the nose is pointing to home, then the rescue must inevitably fail.  If we apply enough pitch to overcome the wind, we have a chance; but if the IMU is incorrect and wants to fly us away from home, then when we apply that strong pitch forward, we will go extremely fast downwind.  True, the IMU will correct, at least partially starting to turn, but the arc to return is very long and very slow, and not until the nose is effectively close to pointing to home can the pitch angle succeed in making any headway against the wind.  In practice, this is why I lost a quad of mine to a sanity check recently while testing this exact situation.  So we kind of need that angle to provide speed into the wind, but adding the same angle when the nose is not true to home will be disastrous also.  By constraining the angle, especially when the velocity away is high, then we don't fly at high speed down-wind, and we don't make such a long arc when the IMU corrects, but then we don't have the forward angle to defeat the wind.

So overall I think it is simply impossible to handle a certain amount of wind, with GPS data alone.  The IMU will always be overwhelmed at some wind velocity, and without accurate knowledge of the true attitude of the quad, eg from a Mag, such rescues are doomed.

So this PR should make it so that the quad will handle some degree of wind drift, and previous IMU disorientation in reasonable air conditions, effectively.  For windy conditions it should still work, to a point, even if the return is a bit all over the place.  And we would need to accept that in strong wind, there is no chance, without good Mag code working properly.  If we have a Mag, it's easy, point the nose to home, lean forward, no problem.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -415,7 +415,14 @@ static float imuCalcKpGain(timeUs_t currentTimeUs, bool useAcc, float *gyroAvera
+        ret = imuRuntimeConfig.dcm_kp;
+#ifdef USE_GPS_RESCUE
+        if (FLIGHT_MODE(GPS_RESCUE_MODE)) {
+            ret *= gpsRescueGetDcmKpModifier();
```
Simply replace `bool useCOG` with `float useCOG`. 0 - ignore COG, 1.0 - current behavior, 2 - 2x faster convergence (yaw gain).

ctzsnooze replied:
OK, however if I bring `useCOG` in as a float, as you say above, range say 0.0 to 3.0, what should be multiplied by that float to action only a change in gain for yaw?  Sorry to be dumb.

ledvinap replied:
https://github.com/betaflight/betaflight/blob/master/src/main/flight/imu.c#L204

`e[xyz]` is error quaternion, this error is minimized. For COG, all that is necessary is to scale `ez_ef`

ctzsnooze replied:
Thank you @ledvinap for that clarification.  Yaw only will be better, I think; I'll try this method.

PS @jsk2084 tested limiting the the pitch angle constraint and dcm_kp modification factor to 3, not 5.  This allowed more pitch angle, and we got more speed, but the heading was 'loose' and 'slow'.   We will test to see if more `Kp` gain with that less aggressive pitch angle reduction is better, or worse.

ledvinap commented:
IMO it is worth trying yaw-only gain staging during rescue. Implementation is quite trivial, I'll look into it.
In long term, better IMU algorithm, possibly even with wind estimation, will solve most of the problems (= some form of Kalman estimator ;-) ). It shouldn't be that hard to port it from other FC (but it will probably require dual sample rate to support high gyro rates. That may complicate thing a bit)

ledvinap commented:
IMU heading align:

1. finish altitude phase (roll/pitch setpoint zero; yaw to home; cog disabled)
   - possibly pre-estimate drift from GPS change in last phase of ascent

2. roll/pitch setpoint zero; yaw fixed; altitude fixed; cog disabled
   - wait until timeout (few s) or significant GPS change (several meters, dependent on GPS resolution)
   - estimate velocity vector during this phase (both wind and ACC misalignment is estimated here)

3.  roll=0; pitch = small forward angle; yaw fixed; altitude fixed; cog disabled
    - wait until significant GPS change or safety timeout
       - subtract velocity 2 when detecting 'significant' change; this will improve error offset/wind handling
    - estimate velocity vector

4. Get forward vector (velocity 3. - velocity 2.)
   - set IMU yaw from forward vector
   - (maybe) set yaw to home, wait until error is reasonable
   - (maybe) set RTH velocity setpoint
   - enable COG

4. Continue RTH phases

ctzsnooze commented:
Thanks @ledvinap for your thoughts, above.  Indeed it would be awesome to look at a more comprehensive, and less empirical, solution. 

For now, if we can make it work nicely in light wind, and be quite OK up to say 30-40kph winds, that would be good.   For higher wind speeds the probability of failure is likely to increase.  Maybe we then advise users who need reliable returns above a certain wind velocity to use a Magnetometer.  That's about the best I'm hoping for, at this point.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -219,7 +219,15 @@ static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
-        const float ez_ef = (- sin_approx(courseOverGround) * rMat[0][0] - cos_approx(courseOverGround) * rMat[1][0]);
+        float ez_ef = (- sin_approx(courseOverGround) * rMat[0][0] - cos_approx(courseOverGround) * rMat[1][0]);
+
+#ifdef USE_GPS_RESCUE
```
(good enough for testing)
RESCUE related code should not be here. Replace `useCog` with gain and move this code to `imuMahonyAHRSupdate` call place.

ctzsnooze replied:
(just tested, handled 'normal' rescues in still air with moderate IMU disorientation very well)

I'll try to do as you say, sorry for not getting it right, but I am not much good as a coder.

ledvinap replied:
It's just cosmetics - keep IMU code as encapsulated as possible

ctzsnooze replied:
I looked at trying to use your suggestion to make `useCOG` as a float, but I see that `useCOG` is a logical operator in several places.
In particular I've never quite followed the logic of these lines:
https://github.com/betaflight/betaflight/pull/12738/files#diff-de6b2b61376cbb30512686ebb362260facb69cf9a4281ebdcd59af61c0306f7aL500-L506
I wouldn't want to mess this up.

ledvinap replied:
What about `useCOG ? imuCogGain : 0.0f`. And in Mahony update, rename parameter to `float cogGain`, but still test for zero (to skip calculation and avoid unlikely case when  CoG is NaN)

ctzsnooze replied:
If it's OK I would like to look at this code idea in a separate PR.  
I say that because we have this arbitrary 2.0m/s on/off behaviour of the mahony update.  I am wondering if this should be re-considered.   Perhaps it could be a gain factor rather than an on/off method.
I would like to look at refactoring the useCog and consider related issues later because this PR is functionally solid now, and it may take a while to do the useCog refactoring.

ledvinap replied:
> Perhaps it could be a gain factor rather than an on/off method
You still want zero gain (no CoG) on low speed. But you can ramp gain when minimum speed is exceeded.

Modifying CoG gain inside inner IMU code is ugly. It should be refactored ASAP. Only few lines of code is needed.

ctzsnooze replied:
@ledvinap I converted `useCOG` to a float `cogYawGain` in commit [9af3cb8ae](https://github.com/betaflight/betaflight/pull/12738/commits/9af3cb8ae5aa5747cd7ee95d453a37b31a0021d2), and it seems to work very well.  Thanks for encouraging me to do this.
Also I increased the rate of acquisition of gain by a factor of 2x, and tested it extensively.  I didn't have a lot of wind, but did test with 180 degree IMU errors and a wide range of velocities, including >120kph away from home, and it worked better than before.  The IMU correction was quicker, obviously, leading to less overshoot and tighter radius corrections.  I was concerned that it might rotate excessively with severe errors but I think the gain was at the limit of 6x on previous tests anyway, and in any case there were some rescues where the IMU gain resulted in an overcorrection, but this was quickly resolved.  

The worst performing situation now is where there is 180 degree IMU disorientation and the quad is 'dead still' in the air at the time of the rescue.  This is below the 2m/s threshold for IMU adjustment.  It must get more than 2m/s before the IMU will even consider the groundspeed.  With no groundspeed to start with, quad must pitch forward to gain velocity.  Since there is little pitch angle attenuation, to start with, it starts to gain quite a bit of speed over time.  With the lag in GPS data, it takes some time before the velocity change results in increase in IMU gain, and by then the velocity is appreciable.  The turn is initially a bit slower than you'd like, and it can seem like it just won't turn at all.  Hence the overshoot distance traveled is further than you'd like, maybe 50m, but once the groundspeed has built up, and is properly detected, the correction becomes fast.  This results in some over-correction, but that is quite good, since it brings you back towards the intended path).  

Hence a user in this circumstance needs about 50m of clear air around them to cover their worst possible outcome.  This is probably a good general recommendation for this code.  At high speed the relative problem seems less but the absolute drift is quite large, just the pre-existing drift speed during the climb results in a lot of travel in those situations.  

In all other cases, having some pre-existing drift when the IMU is 180 degrees wrong seems to make things easier and quicker now; if the drift is away from home, there is already some velocity away, leading to a very quick fix.  If the drift is towards home, but you got the IMU 180 wrong, it will still rotate and fly away from home, but the net effect isn't as bad as a standing start, because we cross the 2m/s 'dead spot' with reasonable velocity already acquired.  Sideways drifts and IMU errors at less than 180 degree all have some groundspeed which helps the IMU to adjust more quickly.

Overall this is pretty good I think.

ledvinap commented:
It would be great to cleanup rescue code state machine ...

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -112,6 +112,8 @@ typedef struct {
     float maxPitchStep;
     float absErrorAngle;
+    float groundspeedErrorRatio;
+    float dcm_kpModifier;
```
ha ha yes, for sure :-)

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +569,21 @@ static void sensorUpdate(void)
     rescueState.sensor.groundSpeedCmS = gpsSol.groundSpeed; // cm/s
 
+    // values to modify dcm_kp to a low value when climbing, and a higher value when there is a velocity error
+    if (gpsRescueConfig()->rescueGroundspeed) {
+        const float setGroundspeedInv = 1.0f / gpsRescueConfig()->rescueGroundspeed;
+        const float groundspeedPitchAttenuator = fabsf(rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) * setGroundspeedInv;
+        // 0 if groundspeed = velocity to home,
+        // 1 if moving sideways at target groundspeed but zero home velocity,
+        // 2 if moving backwards (away from home) at target groundspeed
+        // 4 if moving backwards (away from home) at twice target groundspeed
+        rescueState.sensor.imuYawGain = constrainf(1.0f + groundspeedPitchAttenuator, 1.0f, 6.0f); // up to 6x increase in IMU Yaw gain
+        rescueState.sensor.groundspeedPitchAttenuator = fminf(rescueState.sensor.imuYawGain, 3.0f); // cut pitch angle by up to one third
+        const bool climbOrRotate = (rescueState.phase == RESCUE_ATTAIN_ALT) || (rescueState.phase == RESCUE_ROTATE);
+        rescueState.sensor.imuYawGain = (climbOrRotate) ? 0.0f : rescueState.sensor.imuYawGain;
+        // zero dcm_kp during climb or rotate, to stop IMU error accumulation arising from drift during long climbs
+    }
```
```suggestion
    if (gpsRescueConfig()->rescueGroundspeed) {
        const float rescueGroundspeed= gpsRescueConfig()->rescueGroundspeed;
        const float groundspeedErrorRatio = fmax(0.0f, rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) / rescueGroundspeed;
        // 0 if groundspeed = velocity to home,
        // 1 if moving sideways at target groundspeed but zero home velocity,
        // 2 if moving backwards (away from home) at target groundspeed
        // 4 if moving backwards (away from home) at twice target groundspeed
        if ((rescueState.phase == RESCUE_ATTAIN_ALT) || (rescueState.phase == RESCUE_ROTATE)) {
            // zero IMU CoG yaw during climb or rotate, to stop IMU error accumulation arising from drift during long climbs
            rescueState.sensor.imuYawGain = 0;
        } else {
            // up to 6x increase in CoG IMU Yaw gain
            rescueState.sensor.imuYawGain =  constrainf(1.0f + groundspeedErrorRatio , 1.0f, 6.0f);
        }
        rescueState.sensor.groundspeedPitchAttenuator = 1.0f / fconstrainf(1.0f + groundspeedErrorRatio , 1.0f, 3.0f); // cut pitch angle by up to one third
    }
```

ctzsnooze replied:
Thank you, very nice refactoring.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -650,8 +667,11 @@ void descend(void)
         // if quad drifts further than 2m away from home, should by then have rotated towards home, so pitch is allowed
         rescueState.intent.rollAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.intent.proximityToLandingArea;
         // reduce roll capability when closer to home, none within final 2m
+        rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle / rescueState.sensor.groundspeedPitchAttenuator;
```
```suggestion
        rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.sensor.groundspeedPitchAttenuator;
```
It will save one divide and IMO is easier to understand

ctzsnooze replied:
Yes, thank you, it is definitely better like this.  Very grateful for your suggestions!

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -758,7 +778,8 @@ void gpsRescueUpdate(void)
         }
         if (rescueState.sensor.absErrorAngle < 30.0f) {
-            rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle; // allow pitch
+            // allow pitch, limiting allowed angle if there is a significant groundspeedPitchAttenuator
```
```suggestion
            // allow pitch, limiting allowed angle if we are drifting away from home

ledvinap replied:
quad is leveled (roll,pitch=0) during previous phase. So groundspeed here is drift + wind; unless this phase was entered very quickly.
Value will be changed on next GPS update anyway ...

ctzsnooze replied:
Thank you, again your suggestion is excellent and the comment entirely correct.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -112,6 +112,8 @@ typedef struct {
     float maxPitchStep;
     float absErrorAngle;
+    float groundspeedPitchAttenuator;
+    float imuYawGain;
```
```suggestion
    float imuCoGYawGain;
```
(sorry, missed that before. Only CoG is affected, other sources of Yaw are unchanged (= compass)

ctzsnooze replied:
Nice.  Done.

haslinghuis commented on src/main/flight/gps_rescue.c:
```diff
@@ -792,8 +816,10 @@ void gpsRescueUpdate(void)
         rescueState.intent.rollAngleLimitDeg = 0.5f * rescueState.intent.velocityItermRelax * gpsRescueConfig()->maxRescueAngle;
         // gradually gain roll capability to max of half of max pitch angle
-
+ 
```
:space_invader:

ctzsnooze replied:
thanks

haslinghuis commented on src/main/flight/imu.c:
```diff
@@ -415,10 +423,10 @@ static float imuCalcKpGain(timeUs_t currentTimeUs, bool useAcc, float *gyroAvera
-       }
+        ret = imuRuntimeConfig.dcm_kp;
+        if (!armState) {
+            ret = ret * 10.0f; // Scale the kP to generally converge faster when disarmed.
```
```suggestion
            ret *= 10.0f; // Scale the kP to generally converge faster when disarmed.
```

ctzsnooze replied:
Done

ctzsnooze commented:
Thanks @ledvinap .  Great suggestions, all implemented.
We have tested making the Cog adjustment for yaw only, and it does seem a bit more stable when IMU Gain was boosted. 
Also on testing different values the 'denominator' of the groundspeedError value, which was previously set to the user's configured return speed, it became apparent that this value fundamentally affects the 'slope' of the change in IMU yaw gain.  Low denominator values caused the IMU gain to quickly reach maximum effect even with relatively modest drift.
The intent is to modify the IMU when drift is significant, and to have a range of adjustment that matches the ability of a typical quad to handle the error.
When set around 10, we had good results, so I have locked this value as the denominator.
In a practical sense, an opposing windspeed of 25-30m/s are about the limit of what we can handle at present without a magnetometer.  At some point, just get too much drift downwind, and make too little headway against it, leading to unwanted IMU readjustment from the drift that causes false heading rotation, and when we rotate downwind, it takes time to restore the attitude, and eventually we will get a sanity check failure because we end up spiralling around in circles and never quite breaking into forward flight home.  
In <4.3 code we would have not pitched forward at all, since it targeted groundspeed only, and would always have failed if the drift rate was as low as the target groundspeed, so even though the newer code isn't perfect, it is now a lot better than before.
We should have videos of the behaviour soon.

ledvinap commented:
> Also on testing different values the 'denominator' of the groundspeedError value, which was previously set to the user's configured return speed, it became apparent that this value fundamentally affects the 'slope' of the change in IMU yaw gain. Low denominator values caused the IMU gain to quickly reach maximum effect even with relatively modest drift. The intent is to modify the IMU when drift is significant, and to have a range of adjustment that matches the ability of a typical quad to handle the error. When set around 10, we had good results, so I have locked this value as the denominator. In a practical sense, an opposing windspeed of 25-30m/s are about the limit of what we can handle at present without a magnetometer.

That's problem with ad-hoc heuristics .. It may also make sense to set speed target only to limited value of user speed for first phase of fly home - speed PID setpoint will be more reasonable during IMU CoG alignment and pich limit may not be necessary (and related I term windup will be less problematic). 

> At some point, we just get too much drift downwind, and make too little headway against it, leading to continual, unwanted, and unhelpful IMU readjustment from the drift that causes false heading rotation. Every time we rotate downwind, our pitch forward, despite being reduced, adds to windspeed, and the quad goes fast downwind; although the higher IMU gain turns quickly back into wind, and we accelerate and the IMU gain falls, giving us more time to correct, at some point we just can't win, and eventually we will get a sanity check failure because we end up spiralling around in circles and never quite breaking into clean forward flight home. This seems to happen around 25-30m/s headwinds. In <4.3 code we would have failed in headwinds, since it targeted groundspeed only, and would always set zero pitch if the drift rate was faster than the target groundspeed. Even though the newer code isn't perfect, it is a lot better than before. We should have videos of the behaviour soon.

Heading acquisition phase described in previous comments may help. But simply using compass in such extreme conditions is best option.
Increasing COG gain to fixed value may also help. Add new phase, beginning of fly home. Terminate when CoG course and heading match (+ small timeout).

ctzsnooze commented:
>Increasing COG gain to fixed value may also help (@ledvinap)

I agree, but being heuristic I have to test it a bit.  I was thinking that when the IMU is sufficiently bad that the quad is under IMU control and flying the wrong way, then we might as well boost the IMU updating rate to the full 6x increase, rather than waiting for it to acquire the full velocity error before doing so.  At this point I'd like to confirm how robust the present solution is.

There is a potential issue with an 'all or none' IMU boost.

Imagine that the IMU is not disoriented at all, and the quad rotates perfectly and points to home.  But also that it has strong rearwards drift, from wind or prior forward speed.  In this case, the PR code will constrain the pitch angle, but the quad will pitch forward.  Our problem here is that it will take time to overcome the drift.  During this time, the IMU sees drift away from home and would start to develop a 180 degree yaw error if it had enough time to do so.  If we boost the IMU yaw gain strongly even with relatively little groundspeed error, then during the time it took to accelerate, the IMU could go 180 degrees wrong and rotate the quad to be pointing away.  

Hence I think we need this linear ramp in IMU gain that is proportional to the groundspeed error. In high wind settings, the IMU gain must generally be slower when the error is low, and faster when the error is high.  If we were to be fast always, which I tested previously, there is a risk, in high wind, of spiralling around with the IMU gain always high, with relatively constant pitch angle, and the IMU always wrong.  With the current code we rely on the upwind part of the spiral gaining velocity to home, reducing groundspeed error, and thereby gaining more pitch and lowering the IMU gain, relative to a down-wind leg.  This hopefully will be sufficient for an 'upwind' leg to eventually dominate and fly the quad home.

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -112,6 +112,8 @@ typedef struct {
     float alitutudeStepCm;
     float maxPitchStep;
     float absErrorAngle;
+    float groundspeedErrorRatio;
```
done

KarateBrot commented on src/main/common/maths.h:
```diff
@@ -34,6 +34,8 @@
 
 // Use floating point M_PI instead explicitly.
 #define M_PIf       3.14159265358979323846f
+#define M_TWO_PIf   6.28318530717958647693f
+#define M_HALF_PIf  1.57079632679489661923f
```
I'm pretty sure you don't need to do that as the compiler already replaces terms like `2 * M_PIf` with the resulting constant of `6.28318530717958647693f`.

So you can remove that and just use terms like `2 * M_PIf` and `M_PIf / 2` for better readability.

ledvinap replied:
(compiler is handling constant expressions, not preprocessor. But result is the same)

ctzsnooze replied:
I removed all this refactoring and will make a separate PR for it.
I only did it here because I noticed it.
We use sinApprox quite a lot, especially with the earth referencing of Angle mode and here in a Rescue, on multiple axes, so optimising this function for speed over readability is desirable.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -453,7 +456,7 @@ static void performSanityChecks(void)
         const float velocityToHomeCmS = previousDistanceToHomeCm - rescueState.sensor.distanceToHomeCm; // cm/s
         previousDistanceToHomeCm = rescueState.sensor.distanceToHomeCm;
-        rescueState.intent.secondsFailing += (velocityToHomeCmS < 0.5f * rescueState.intent.targetVelocityCmS) ? 1 : -1;
+        rescueState.intent.secondsFailing += (velocityToHomeCmS < 0.1f * rescueState.intent.targetVelocityCmS) ? 1 : -1;
```
Is this tested? 10% margin feels quite small ...

ctzsnooze replied:
It's still bringing the quad home :-)
The argument here is that so long as we are heading towards home, we should reduce the counter that brings us closer to disarming, even if we only make 10% of the requested headway, it's better than nothing.
In practicality, most rescues in 'normal' situation achieve very close to requested return velocities.
However, there are some unusual situations where this may not be the case.  For instance, if someone configured a fast return rate, but stupidly set a low value for the maximum pitch angle, or lacked adequate max throttle, and then encountered a headwind, their return to home velocity as a percentage of the set return velocity could be quite low, even though they could be making steady headway.
If they fail to make any headway (stuck in tree) or if they are consistently moving negative (away) then the sanity check will still work well.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -263,7 +263,7 @@ static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
     float recipAccNorm = sq(ax) + sq(ay) + sq(az);
     if (useAcc && recipAccNorm > 0.01f) {
-        // Normalise accelerometer measurement
+        // Normalise accelerometer measurement when smoothed acc is within 20% of 1G
```
20% margin is tested elsewhere. This only prevents division by zero.

ctzsnooze replied:
True, but useAcc is only true when all smoothed axes are within 20% of 1G.
I'll modify the comment, but I wanted it to let the reader know that the code would not run unless the smoothed acc value on every axis is within 20% of 1G, ie the quad is stable.

ctzsnooze commented:
@ledvinap @KarateBrot... I accepted your comments above and:
- removed the refactoring involving PI, 
- ensured that the IMU gain is only modified when we are in the GPS Rescue flight mode
- modified a comment

Both Sek101 and I have tested this code now, and it works very well.  The higher gain for the IMU boost factor helps deal with IMU errors more quickly when needed, and does not induce circling or excessive overshoot, even with 120kph backwards flight with a full 180 degree IMU error.  I think the gain factor is about as high as it should be.

I'm very grateful for your advice and assistance with this PR.

A final check, and if agreeable, approval, would be much appreciated.  I am confident that it will make GPS rescues much safer.

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -495,14 +493,15 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
-
+        if (FLIGHT_MODE(GPS_RESCUE_MODE)) {
+            cogYawGain = gpsRescueGetImuYawCogGain(); // do not modify IMU yaw gain unless in a rescue
+        }
```
```suggestion
        } else {
            cogYawGain = 1.0f;
        }
```

CoG is used to correct yaw in forward flight. Ideally, thrust vector horizontal vector shall be computed from roll/pitch and used to adjust `courseOverGround`. But for now, don't change existing behavior.

ctzsnooze replied:
Here I think the alternative,or `else` condition, on line 498, should be a zero value, not a 1, to return the float equivalent of the boolean OFF.  That's why I initialise `cogYawGain` to `0.0f` on line 477.  Our intent is to do nothing unless we are in `FLIGHT_MODE(GPS_RESCUE_MODE)`.  If we keep `cogYawGain` at zero, it won't pass the check for `if (cogYawGain != 0.0f)` back on line 213.

I agree it would be great to get a vector from roll/pitch attitude to help evaluate courseOverGround, especially if integrated over time, and also agree that for now I'd not want to change the existing behaviour.

ctzsnooze replied:
@ledvinap Looks like this was merged before I had time to resolve this.

I'll make a small PR to restore the original behaviour.  I'm not 100% exactly sure what is best, but I guess we should correct the quad's attitude throughout the flight, since otherwise it could drift.  And in any case I had no intention of altering the existing behaviour.

Also I was reflecting on your comment, earlier, about using Pitch angle as a factor for IMU gain.

This could be done quite easily in the GPS Rescue case, and would be beneficial in the situation of a rescue initiated with very low groundspeed but an IMU that is 180 degrees wrong.  Currently the quad will pitch forward, but it takes time to accumulate velocity before the IMU gain for yaw is increased.  If we had a signal to boost IMU directly related to forward pitch angle (before adjustment for groundspeed error), this is nearly always high to maximal in groundspeed error settings, and is high from the moment the rescue starts.  Using positive pitch angle to generate IMU yaw gain would act as a leading indicator of a 180 degree IMU error.  Also, generally speaking, it is more likely that if the pitch angle is positive, then we should have direction over ground that corresponds 'better than usually' to the orientation of the nose of the quad.

A flight home at high angle against a strong headwind, or with a high target velocity, could require a high pitch angle all the time.  The quad would then fly home with an increased IMU gain on yaw the whole time.  In both these situations we should be pointing fairly directly to home, and the IMU error in any case resolves to reduce the effect of sideways drag.  IMU yaw gain is functionally a time constant, so we are not going to change the magnitude of anything, just how quickly it happens, and this is probably not so bad in higher pitch angle environments.  

I'm inclined to test this out.  It will require some heuristic gain factor though :-)

Thanks again for all your input to date, I very much appreciate everything you've done here.

ctzsnooze commented on src/main/flight/imu.c:
```diff
@@ -476,8 +474,8 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
     bool useMag = false;
-    bool useCOG = false; // Whether or not correct yaw via imuMahonyAHRSupdate from our ground course
-    float courseOverGround = 0; // To be used when useCOG is true.  Stored in Radians
+    float cogYawGain = 0.0f; // IMU yaw gain to be applied in imuMahonyAHRSupdate from ground course, default to no modification
```
OK, that's more clear

ctzsnooze commented on src/main/flight/imu.c:
```diff
@@ -210,17 +210,15 @@ static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
     float ex = 0, ey = 0, ez = 0;
-    if (useCOG) {
+    if (cogYawGain != 0.0f) {
+        // if in a GPS Rescue, boost IMU yaw gain when course over ground and velocity to home differ significantly
```
Thanks, will do

ctzsnooze commented on src/main/flight/gps_rescue.c:
```diff
@@ -567,6 +569,25 @@ static void sensorUpdate(void)
 
+    // when there is significant velocity error, increase IMU COG Gain for yaw to a higher value, and reduce max pitch angle
+    if (gpsRescueConfig()->rescueGroundspeed) {
+        const float rescueGroundspeed = 1000.0f; // equivalent of fixed 10 m/s groundspeed
```
OK, I'll make that change.

ledvinap commented on src/main/flight/gps_rescue.c:
```diff
@@ -648,10 +669,13 @@ void descend(void)
         // reduce target velocity as we get closer to home. Zero within 2m of home, reducing risk of overshooting.
         // if quad drifts further than 2m away from home, should by then have rotated towards home, so pitch is allowed
-        rescueState.intent.rollAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.intent.proximityToLandingArea;
+        rescueState.intent.rollAngleLimitDeg = 0.5f * gpsRescueConfig()->maxRescueAngle * rescueState.intent.proximityToLandingArea;
```
I think code comment is enough. As you say, the 50% limit is more than enough. But make it easier for programmers to understand it and avoid confusion.

ctzsnooze replied:
Will do this in separate PR

ledvinap commented on src/main/flight/imu.c:
```diff
@@ -201,7 +201,7 @@ static float invSqrt(float x)
                                 bool useAcc, float ax, float ay, float az,
                                 bool useMag,
-                                bool useCOG, float courseOverGround, const float dcmKpGain)
+                                float cogYawGain, float courseOverGround, const float dcmKpGain)
```
IIRC the error quaternion is rotation in around UP direction in earth frame, so it should affect heading only. Definition of yaw is a bit more complicated (nonzero roll/pitch)

But the gain is more related to CoG parameter of IMU

ctzsnooze replied:
Good points.  

If its OK I'll keep the name the same everywhere, since it makes 'searching' for this particular parameter/concept a bit easier when looking for places in the code where we are dealing with it.

haslinghuis commented:
@ledvinap please let me know if ready to merge

sugaarK commented:
I flew a number of packs with this yesterday and it works VERY well one warning is not to stand to close to your launch point as it’ll try and land on your head 😂

sugaarK commented:
I flew a number of packs with this yesterday and it works VERY well one warning is not to stand to close to your launch point as it’ll try and land on your head 😂

sugaarK reviewed (approved):
I flew a number of packs with this yesterday and it works VERY well one warning is not to stand to close to your launch point as it’ll try and land on your head 😂

ctzsnooze commented:
@ledvinap I've tried to address your remaining comments in https://github.com/betaflight/betaflight/pull/12789
